chainer/functions/math/max.py

```diff
@@ -0,0 +1,111 @@
+from chainer import cuda
+from chainer import function
+from chainer.utils import type_check
+
+
+class SelectorBase(function.Function):
+    """Select an array element from a given axis or set of axes."""
+
+    def __init__(self, axis=None, keepdims=False):
+        self.keepdims = keepdims
+        if axis is None:
+            self.axis = None
+        elif isinstance(axis, int):
+            self.axis = (axis,)
+        elif isinstance(axis, tuple) and all(isinstance(a, int) for a in axis):
+            if len(set(axis)) != len(axis):
+                raise ValueError('duplicate value in axis: ({})'.format(
+                    ', '.join(map(str, axis))))
+            self.axis = axis
+        else:
+            raise TypeError('None, int or tuple of int are required')
+
+    def _fwd(self, x, xp):
+        raise NotImplementedError('_fwd should be implemented in sub-class.')
+
+    def check_type_forward(self, in_types):
+        type_check.expect(
+            in_types.size() == 1,
+            in_types[0].dtype.kind == 'f'
+        )
+
+        if self.axis is not None:
+            for axis in self.axis:
+                if axis >= 0:
+                    type_check.expect(
+                        axis < in_types[0].ndim,
+                    )
+                else:
+                    type_check.expect(
+                        -axis - 1 < in_types[0].ndim,
+                    )
+
+    def forward(self, x):
+        xp = cuda.get_array_module(*x)
+        self.y = xp.asarray(self._fwd(x[0], xp))
+        return self.y,
+
+    def backward(self, x, gy):
+        x = x[0]
+
+        if self.axis is None:
+            axis = range(x.ndim)
+        else:
+            axis = [ax % x.ndim for ax in self.axis]
+
+        # Add broadcastable dimensions to y and gy
+        # for each one that was reduced in the forward operation
+        shape = [s if ax not in axis else 1 for ax, s in enumerate(x.shape)]
+        gy = gy[0].reshape(shape)
+        y = self.y.reshape(shape)
+
+        # Compute the gradient
+        return gy * (x == y),
+
+
+class Max(SelectorBase):
+
+    def __init__(self, axis=None, keepdims=False):
+        super(Max, self).__init__(axis, keepdims)
```

Comment:
You can remove this `__init__` since it only calls `__init__` of super class. Base class's `__init__` is automatically called if derived class does not implement its own `__init__` method.